spec/integration/rails/finders_spec.rb

```diff
@@ -383,58 +393,58 @@ def json
-        #}
-        #house = json_includes('houses')[0]['attributes']
-        #expect(house['name']).to be_present
-        #expect(house['house_description']).to be_present
-        #expect(house['house_price']).to eq(1_000_000)
-        #condo = json_includes('condos')[0]['attributes']
-        #expect(condo['name']).to be_present
-        #expect(condo['condo_description']).to be_present
-        #expect(condo['condo_price']).to eq(500_000)
-      #end
-
-      #it 'allows sparse fieldsets for the sideloaded resource' do
-        #get :index, params: {
-          #include: 'dwelling',
-          #fields: { houses: 'name', condos: 'condo_description' }
-        #}
-        #house = json_includes('houses')[0]['attributes']
-        #expect(house['name']).to be_present
-        #expect(house).to_not have_key('house_description')
-        #expect(house).to_not have_key('house_price')
-        #condo = json_includes('condos')[0]['attributes']
-        #expect(condo['condo_description']).to be_present
-        #expect(condo).to_not have_key('name')
-        #expect(condo).to_not have_key('condo_price')
-      #end
-
-      #it 'allows extra fields and sparse fieldsets for the sideloaded resource' do
-        #get :index, params: {
-          #include: 'dwelling',
-          #fields: { houses: 'name', condos: 'condo_description' },
-          #extra_fields: { houses: 'house_price', condos: 'condo_price' }
-        #}
-        #house = json_includes('houses')[0]['attributes']
-        #condo = json_includes('condos')[0]['attributes']
-        #expect(house).to have_key('name')
-        #expect(house).to have_key('house_price')
-        #expect(house).to_not have_key('house_description')
-        #expect(condo).to have_key('condo_description')
-        #expect(condo).to have_key('condo_price')
-        #expect(condo).to_not have_key('name')
-      #end
-
-      #it 'allows additional levels of nesting' do
-        #get :index, params: { include: 'dwelling.state' }
-        #expect(json_includes('states').length).to eq(1)
-      #end
-    #end
+    context 'sideloading polymorphic belongs_to' do
+      it 'allows extra fields for the sideloaded resource' do
+        get :index, params: {
+          include: 'dwelling',
+          extra_fields: { houses: 'house_price', condos: 'condo_price' }
+        }
+        house = json_includes('houses')[0]['attributes']
+        expect(house['name']).to be_present
+        expect(house['house_description']).to be_present
+        expect(house['house_price']).to eq(1_000_000)
+        condo = json_includes('condos')[0]['attributes']
+        expect(condo['name']).to be_present
+        expect(condo['condo_description']).to be_present
+        expect(condo['condo_price']).to eq(500_000)
+      end
+
+      it 'allows sparse fieldsets for the sideloaded resource' do
+        get :index, params: {
+          include: 'dwelling',
+          fields: { houses: 'name', condos: 'condo_description' }
```

Comment:
What about if I have `visas` and `mastercards`, which are both `credit_cards` subclasses? Could I say `fields: { credit_cards: 'number,billing_zip' }` here?  I think this is pretty important behavior.

On a related note that I’ve only just considered: what happens if I’m a client who doesn’t care about subclasses? I know everything quacks like a credit card and that’s the resolution I care about. Our current system doesn’t provide any way to declare a model for type “credit_cards” and have everything that comes back as a subclass type get instantiated correctly. This might not be addressable without enumerating all possible child types in the client without a response spec change (“ancestor_types” field or something).

Reply:
Good point. Currently you can use `credit_cards` in the query for everything but fields - there we have a spec limitation that `fields` has to align with `type`...a limitation we should break. I'll add this to my list of notes.

> Our current system doesn’t provide any way to declare a model for type “credit_cards” and have everything that comes back as a subclass type get instantiated correctly. 

We actually [do have this behavior](https://github.com/richmolj/jsonapi_compliable/blob/poly_belongs_to/spec/polymorphism_spec.rb#L21). So you could filter on `credit_cards` and get back `visas`/etc with appropriate serialization. Fieldsets is a notable exception we should handle - the *request* should always be about the parent, the *response* is about the children.

I think you're making a larger point that I think I agree with, but would need more specific examples.

Reply:
Oh I know I can query on the supertype. My point was that if my models were:

``` ruby
class CreditCard
  attribute :number
  attribute :billing_zip
end

class Visa < CreditCard
  attribute :visa_support_phone
end

class Mastercard < CreditCard
  attribute :mastercard_support_phone
end
```

If on the JSORM side, I only needed:

``` typescript
class CreditCard extends ApplicationRecord {
  static jsonapiType = 'credit_cards'

  @Attr number : string
  @Attr billingZip : number
}
```

There currently isn't enough information in the response for `CreditCard.all()` to return a list of `CreditCard` instances. I'd have to either declare children I didn't need or enumerate the types it can be instantiated from.

Reply:
Right, it's a requirement to define the subclasses. I think if the server cared enough to change the jsonapi `type`, you should care enough to define a subclass (which can just inherit `CreditCard`). I'm open to changing my mind but that is the thought.

Reply:
I disagree a bit in principle, but from a Moreno practical perspective, say in my above example a consumer does that work and adds their Visa and MasterCard classes which don’t do anything but inherit the parent and define the child type. I now have a UI that reliably lists all user credit cards. If we start supporting Amex tomorrow and push that new version, my client, who only care about listing all of a user’s credit cards regardless of type is now not listing them all because they didn’t add an extra model because they didn’t know I pushed a new subclass.

Another case around querying that needs resolving: if I do Visa.all(), should the convention be to ask the credit_cards resource with a type filter? I would think so but we don’t currently support this so would need to decide on that.

Reply:
These are great points. I think we'll have to tackle this when we get to the schema/jsorm portion, but I buy the argument that the client shouldn't have to specify subclasses.

>  if I do Visa.all(), should the convention be to ask the credit_cards resource with a type filter? 

I think we will need to support this, with an easy way to go to `/visas` instead.